lib/Modules/HyperVCIHooks/HyperVCIHooks.psm1

```diff
@@ -0,0 +1,1538 @@
+
+    $packageFormat = $URL.Split('.')[-1]
+    $acceptedFormats = @('msi', 'exe')
+    if ($packageFormat -notin $acceptedFormats) {
+        Throw ("Cannot install the package found at this URL: $URL " +
+               "Unsupported installer format.")
+    }
+
+    $installerPath = Get-PackagePath $URL $Checksum $HashingAlgorithm
+    $stat = Start-Process -FilePath $installerPath -ArgumentList $ArgumentList `
+                          -PassThru -Wait
+    if ($stat.ExitCode -ne 0) {
+        throw "Package '$URL' failed to install."
+    }
+    Remove-Item $installerPath
+
+    Write-JujuLog "Finished installing package."
+}
+
+
+function Start-GitClonePull {
+    Param(
+        [string]$Path,
+        [string]$URL,
+        [string]$Branch="master"
+    )
+
+    if (!(Test-Path -Path $Path)) {
+        Start-ExecuteWithRetry {
+            Start-ExternalCommand -ScriptBlock { git clone $URL $Path } `
+                                  -ErrorMessage "Git clone failed"
+        }
+        pushd $Path
+        Start-ExternalCommand -ScriptBlock { git checkout $Branch } `
+                              -ErrorMessage "Git checkout failed"
+        popd
+    } else {
+        pushd $Path
+        try {
+            $gitPath = Join-Path $Path ".git"
+            if (!(Test-Path -Path $gitPath)) {
+                Remove-Item -Recurse -Force *
+                Start-ExecuteWithRetry {
+                    Start-ExternalCommand -ScriptBlock { git clone $URL $Path } `
+                                          -ErrorMessage "Git clone failed"
+                }
+            } else {
+                Start-ExecuteWithRetry {
+                    Start-ExternalCommand -ScriptBlock { git fetch --all } `
+                                          -ErrorMessage "Git fetch failed"
+                }
+            }
+            Start-ExecuteWithRetry {
+                Start-ExternalCommand -ScriptBlock { git checkout $Branch } `
+                                      -ErrorMessage "Git checkout failed"
+            }
+            Get-ChildItem . -Include *.pyc -Recurse | foreach ($_) { Remove-Item $_.fullname }
+            Start-ExternalCommand -ScriptBlock { git reset --hard } `
+                                  -ErrorMessage "Git reset failed"
+            Start-ExternalCommand -ScriptBlock { git clean -f -d } `
+                                  -ErrorMessage "Git clean failed"
+            Start-ExecuteWithRetry {
+                Start-ExternalCommand -ScriptBlock { git pull } `
+                                      -ErrorMessage "Git pull failed"
+            }
+        } finally {
+            popd
+        }
+    }
+}
+
+
+function Install-OpenStackProjectFromRepo {
+    Param(
+        [string]$ProjectPath
+    )
+
+    Start-ExternalCommand -ScriptBlock { pip install -e $ProjectPath } `
+                          -ErrorMessage "Failed to install $ProjectPath from repo."
+}
+
+
+function Start-GerritGitPrep {
```

Comment:
I would definitely investigate aggregating some of those Start-ExternalCommands into one. See above comments.